clippy_lints/src/matches.rs

```diff
@@ -1553,47 +1558,199 @@ where
 
 mod redundant_pattern_match {
     use super::REDUNDANT_PATTERN_MATCHING;
-    use crate::utils::{match_qpath, match_trait_method, paths, snippet, span_lint_and_then};
+    use crate::utils::{
+        implements_trait, is_type_diagnostic_item, is_type_lang_item, match_qpath, match_trait_method, match_type,
+        paths, snippet, snippet_with_applicability, span_lint_and_then,
+    };
     use if_chain::if_chain;
     use rustc_ast::ast::LitKind;
     use rustc_errors::Applicability;
-    use rustc_hir::{Arm, Expr, ExprKind, MatchSource, PatKind, QPath};
+    use rustc_hir::{
+        intravisit::{walk_expr, ErasedMap, NestedVisitorMap, Visitor},
+        Arm, Block, Expr, ExprKind, LangItem, MatchSource, Node, PatKind, QPath,
+    };
     use rustc_lint::LateContext;
+    use rustc_middle::ty::{self, subst::GenericArgKind, Ty};
     use rustc_span::sym;
 
     pub fn check<'tcx>(cx: &LateContext<'tcx>, expr: &'tcx Expr<'_>) {
         if let ExprKind::Match(op, arms, ref match_source) = &expr.kind {
             match match_source {
                 MatchSource::Normal => find_sugg_for_match(cx, expr, op, arms),
-                MatchSource::IfLetDesugar { .. } => find_sugg_for_if_let(cx, expr, op, arms, "if"),
-                MatchSource::WhileLetDesugar => find_sugg_for_if_let(cx, expr, op, arms, "while"),
+                MatchSource::IfLetDesugar { contains_else_clause } => {
+                    find_sugg_for_if_let(cx, expr, op, &arms[0], "if", *contains_else_clause)
+                },
+                MatchSource::WhileLetDesugar => find_sugg_for_if_let(cx, expr, op, &arms[0], "while", false),
                 _ => {},
             }
         }
     }
 
+    // Check if the drop order for a type matters
+    fn type_needs_ordered_drop(cx: &LateContext<'tcx>, ty: Ty<'tcx>) -> bool {
+        if !ty.needs_drop(cx.tcx, cx.param_env) {
+            false
+        } else if cx
```

Comment:
Why is `needs_drop` not enough here? The [documentation](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_middle/ty/struct.TyS.html#method.needs_drop) says:

> if it returns `false`, then `ty` definitely has no destructor (i.e., no drop glue).

Skimming the query code, it also checks ADT fields and so on.

Reply:
It's not enough to determine whether drop order matters. The drop order for something `Vec` doesn't impact program correctness* and, by extension, neither would `(Vec, Vec)`. `needs_drop` is a good way to short circuit a search through all fields as the query would have already run for every type, and a good chunk of types don't actually need to be dropped.

The whole point of this change is to note that the suggested change is not correct in the case where drop order is being relied on. There's no way to handle this in the general case, but there are a few std types which implement drop only to deallocate memory. 

* I guess you could use `Vec` while implementing the allocator it uses, but that's a rather pathological case.

Reply:
Thanks for clarifying this. Can you add this as a comment to the function, please?